src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs

```diff
@@ -0,0 +1,116 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Logging;
+    using Transport;
+
+    class RecoverabilityExecutor
+    {
+        public RecoverabilityExecutor(IRecoverabilityPolicy recoverabilityPolicy, IEventAggregator eventAggregator, DelayedRetryExecutor delayedRetryExecutor, MoveToErrorsExecutor moveToErrorsExecutor, bool transactionsOn)
+        {
+            this.recoverabilityPolicy = recoverabilityPolicy;
+            this.eventAggregator = eventAggregator;
+            this.delayedRetryExecutor = delayedRetryExecutor;
+            this.moveToErrorsExecutor = moveToErrorsExecutor;
+            this.transactionsOn = transactionsOn;
+
+            raiseNotifications = recoverabilityPolicy.RaiseNotifications;
+            delayedRetriesEnabled = delayedRetryExecutor != null;
+        }
+
+        public Task<ErrorHandleResult> Invoke(ErrorContext errorContext)
+        {
+            if (transactionsOn == false || errorContext.Exception is MessageDeserializationException)
+            {
+                return MoveToError(errorContext);
+            }
+
+            return PerformRecoverabilityAction(errorContext);
+        }
+
+        Task<ErrorHandleResult> PerformRecoverabilityAction(ErrorContext errorContext)
+        {
+            var recoveryAction = recoverabilityPolicy.Invoke(errorContext);
+
+            if (recoveryAction is ImmediateRetry)
+            {
+                return RaiseImmediateRetryNotifications(errorContext);
+            }
+
+            // When we can't do delayed retries then just fall through to error.
+            if (recoveryAction is DelayedRetry && !delayedRetriesEnabled)
+            {
+                Logger.Warn("Current recoverability policy requested delayed retry but delayed delivery is not supported by this endpoint. Consider enabling the timeout manager or use a transport which natively supports delayed delivery. Moving to the error queue instead.");
+                recoveryAction = RecoverabilityAction.MoveToError();
+            }
+
+            var delayedRetryAction = recoveryAction as DelayedRetry;
```

Comment:
I thought the only required change is moving line 49 to 41 and replace `recoveryAction is DelayedRetry ` in the if statement with a null check. that doesn't sound like such a downgrade in readability to me. But maybe I'm also missing something after looking through 248 changed files. Also, it's probably not an impact at all, so if you're happing with the code after thinking through an alternative, I'm good with that too.